File: Server~/src/resources/getAllToolsResource.ts
import { Logger } from '../utils/logger.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { ReadResourceResult } from '@modelcontextprotocol/sdk/types.js';
import { ToolRegistry } from '../tools/base/ToolRegistry.js';
import { zodToReadableSchema } from '../utils/zodToJsonSchema.js';

// Constants for the resource
const resourceName = 'get_all_tools';
const resourceUri = 'unity://all-tools';
const resourceMimeType = 'application/json';

/**
 * Category descriptions mapping (short versions for token efficiency)
 */
const CATEGORY_DESCRIPTIONS: Record<string, string> = {
  'advanced': 'ProBuilder shapes',
  'animation': 'Animations & timeline',
  'asset': 'Asset management',
  'audio': 'Audio sources',
  'build': 'Project building',
  'camera': 'Cameras',
  'component': 'Component operations',
  'components': 'Add components',
  'debug': 'Debugging tools',
  'gameobject': 'GameObject management',
  'lighting': 'Lighting & baking',
  'material': 'Materials & colors',
  'menu': 'Menu execution',
  'meta': 'Meta tools',
  'physics': 'Physics & colliders',
  'prefab': 'Prefabs',
  'scene': 'Scene management',
  'scripting': 'C# scripts',
  'terrain': 'Terrain',
  'testing': 'Testing',
  'ui': 'UI elements',
  'vfx': 'Visual effects'
};

/**
 * Creates and registers the All Tools resource with the MCP server
 * This resource provides ALL tools in ONE request for maximum efficiency
 * 
 * @param server The MCP server instance to register with
 * @param logger The logger instance for diagnostic information
 */
export function registerGetAllToolsResource(server: McpServer, logger: Logger) {
  logger.info(`Registering resource: ${resourceName}`);
      
  // Register this resource with the MCP server
  server.resource(
    resourceName,
    resourceUri,
    {
      description: '⚡ RECOMMENDED! Get ALL tools in ONE request. 🚨 PLAN tool sequence FIRST, THEN execute with discover_and_use_batch. Zero-registration architecture.',
      mimeType: resourceMimeType
    },
    async () => {
      try {
        return await resourceHandler(logger);
      } catch (error) {
        logger.error(`Error handling resource ${resourceName}: ${error}`);
        throw error;
      }
    }
  );
}

/**
 * Handles requests for all tools information
 * Returns ALL tools organized by category in a single response
 * 
 * @param logger The logger instance for diagnostic information
 * @returns A promise that resolves to all tools data
 */
async function resourceHandler(logger: Logger): Promise<ReadResourceResult> {
  // Get all categories from the ToolRegistry
  const categories = ToolRegistry.getCategories();
  const stats = ToolRegistry.getStatistics();
  
  // Build category summary with tool counts only (no individual tools to save tokens)
  const categorySummary: Record<string, { description: string; toolCount: number }> = {};
  let totalToolCount = 0;
  
  // Process each category
  for (const category of categories) {
    const toolClasses = ToolRegistry.getToolsByCategory(category);
    
    if (toolClasses.length === 0) continue;
    
    const toolCount = toolClasses.length;
    totalToolCount += toolCount;
    
    categorySummary[category] = {
      description: CATEGORY_DESCRIPTIONS[category] || 'Tools',
      toolCount: toolCount
    };
  }
  
  // Build response (minimal format to save tokens)
  // Merged from getToolCategoriesResource for simplicity
  const response = {
    _instruction: "🚨 WORKFLOW: 1) Query unity://tool-names/{category} for specific tools, 2) Use discover_and_use_batch for 2+ tools with $.{index}.field chaining",
    totalToolCount: totalToolCount,
    categoryCount: Object.keys(categorySummary).length,
    categories: categorySummary
  };
  
  return {
    contents: [{ 
      uri: resourceUri,
      mimeType: resourceMimeType,
      text: JSON.stringify(response) // No formatting to save tokens
    }]
  };
}

File: Editor/Resources/GetAllToolsResource.cs
using System;
using System.Collections.Generic;
using System.Linq;
using McpUnity.Unity;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace McpUnity.Resources
{
    public class GetAllToolsResource : McpResourceBase
    {
        protected McpUnityServer Server { get; private set; }

        public GetAllToolsResource(McpUnityServer server)
        {
            Server = server;
            Uri = "unity://all-tools";
            Name = "get_all_tools";
            Description = "Get category overview with tool counts (no individual tool details). Use unity://tool-names/{category} to see tools in a category.";
        }

        public override JObject Fetch(JObject parameters)
        {
            try
            {
                var categorySummary = new Dictionary<string, object>();
                var toolCount = 0;

                // Get all tool categories
                var categories = new[]
                {
                    new { name = "scene", desc = "Scene management" },
                    new { name = "gameobject", desc = "GameObject management" },
                    new { name = "physics", desc = "Physics & colliders" },
                    new { name = "ui", desc = "UI elements" },
                    new { name = "scripting", desc = "C# scripts" },
                    new { name = "component", desc = "Component operations" },
                    new { name = "components", desc = "Add components" },
                    new { name = "material", desc = "Materials & colors" },
                    new { name = "camera", desc = "Cameras" },
                    new { name = "lighting", desc = "Lighting & baking" },
                    new { name = "audio", desc = "Audio sources" },
                    new { name = "animation", desc = "Animations & timeline" },
                    new { name = "prefab", desc = "Prefabs" },
                    new { name = "asset", desc = "Asset management" },
                    new { name = "terrain", desc = "Terrain" },
                    new { name = "vfx", desc = "Visual effects" },
                    new { name = "build", desc = "Project building" },
                    new { name = "testing", desc = "Testing" },
                    new { name = "debug", desc = "Debugging tools" },
                    new { name = "menu", desc = "Menu execution" },
                    new { name = "advanced", desc = "ProBuilder shapes" },
                    new { name = "meta", desc = "Meta tools" }
                };

                // Get all registered tools from server
                var allRegisteredTools = Server.GetAllTools();

                // Count tools per category (don't include individual tool details)
                foreach (var category in categories)
                {
                    int categoryToolCount = 0;

                    // Count tools in this category
                    foreach (var tool in allRegisteredTools.Values)
                    {
                        var toolCategory = GetToolCategory(tool.Name);
                        if (toolCategory == category.name)
                        {
                            categoryToolCount++;
                            toolCount++;
                        }
                    }

                    if (categoryToolCount > 0)
                    {
                        categorySummary[category.name] = new
                        {
                            description = category.desc,
                            toolCount = categoryToolCount
                        };
                    }
                }

                var result = new
                {
                    _workflow = "📋 STEP 1/4: Choose category → STEP 2: get_tool_names({category}) → STEP 3: discover_and_use_batch (2+ tools, chain with $.{index}.field)",
                    totalToolCount = toolCount,
                    categoryCount = categorySummary.Count,
                    categories = categorySummary
                };

                return JObject.FromObject(result);
            }
            catch (Exception ex)
            {
                return new JObject
                {
                    ["error"] = true,
                    ["message"] = $"Failed to get all tools: {ex.Message}",
                    ["stackTrace"] = ex.StackTrace
                };
            }
        }

        private string GetToolCategory(string toolName)
        {
            // Scene tools
            if (toolName.Contains("scene") || toolName == "save_scene" || toolName == "add_scenes_to_build")
                return "scene";

            // GameObject tools
            if (toolName.Contains("gameobject") || toolName.Contains("primitive") || toolName.Contains("empty") ||
                toolName == "select_gameobject" || toolName == "update_gameobject" || toolName == "group_gameobjects" ||
                toolName == "align_gameobjects" || toolName == "distribute_gameobjects" || toolName == "replace_gameobjects" ||
                toolName == "batch_rename_gameobjects" || toolName == "cleanup_empty_gameobjects" || toolName == "set_active_state" ||
                toolName == "set_layer" || toolName == "set_tag" || toolName == "copy_transform" || toolName == "snap_to_grid" ||
                toolName == "randomize_transform" || toolName == "delete_gameobject" || toolName == "duplicate_gameobject" ||
                toolName == "set_parent" || toolName == "find_gameobjects")
                return "gameobject";

            // Physics tools
            if (toolName.Contains("rigidbody") || toolName.Contains("collider") || toolName.Contains("physics") ||
                toolName.Contains("navmesh") || toolName.Contains("joint") || toolName.Contains("force") ||
                toolName == "raycast" || toolName == "overlap_sphere")
                return "physics";

            // UI tools
            if (toolName.Contains("ui_") || toolName.Contains("create_event_system") || toolName.Contains("toggle_panel"))
                return "ui";

            // Scripting tools
            if (toolName.Contains("script") || toolName == "recompile_scripts")
                return "scripting";

            // Component tools
            if (toolName.Contains("component") && !toolName.Contains("add_"))
                return "component";

            // Add component tools
            if (toolName.StartsWith("add_") && !toolName.Contains("scene"))
                return "components";

            // Material tools
            if (toolName.Contains("material") || toolName.Contains("color") || toolName.Contains("shader"))
                return "material";

            // Camera tools
            if (toolName.Contains("camera"))
                return "camera";

            // Lighting tools
            if (toolName.Contains("light") || toolName.Contains("bake"))
                return "lighting";

            // Audio tools
            if (toolName.Contains("audio") || toolName.Contains("sound"))
                return "audio";

            // Animation tools
            if (toolName.Contains("animation") || toolName.Contains("animator") || toolName.Contains("timeline"))
                return "animation";

            // Prefab tools
            if (toolName.Contains("prefab"))
                return "prefab";

            // Asset tools
            if (toolName.Contains("asset") || toolName.Contains("import") || toolName.Contains("export"))
                return "asset";

            // Terrain tools
            if (toolName.Contains("terrain"))
                return "terrain";

            // VFX tools
            if (toolName.Contains("vfx") || toolName.Contains("particle"))
                return "vfx";

            // Build tools
            if (toolName.Contains("build"))
                return "build";

            // Testing tools
            if (toolName.Contains("test"))
                return "testing";

            // Debug tools
            if (toolName.Contains("debug") || toolName.Contains("log"))
                return "debug";

            // Menu tools
            if (toolName.Contains("menu"))
                return "menu";

            // Advanced tools
            if (toolName.Contains("probuilder"))
                return "advanced";

            // Meta tools
            if (toolName.Contains("discover_and_use"))
                return "meta";

            return "other";
        }
    }
}
